# Stat 220 Lab 2: Probability Calculation for Red Dragon Inn

### Function A

Calculating the probability of rolling m dice, each with n sides, and having the sum of the dice be greater than x. This calculation should only be implemented for m = 1 through 4 and n is at most 20.

In [25]:
def functionA(m,n,x):
  """
  Calculates the probability of rolling m dice, each with n sides, and having the sum of the dice be greater than x.

  Args:
    m: The number of dice.
    n: The number of sides on each die.
    x: The target sum.

  Returns:
    The probability of the sum of the dice being greater than x.
  """
  # Generate all possible rolls
  rolls = list(itertools.product(range(1, n+1), repeat=m))

  # Calculate the sum of each roll
  sums = np.array([sum(roll) for roll in rolls])

  # Calculate the probability of the sum being greater than x
  return np.mean(sums > x)

**Test Cases**

In [26]:
functionA(1,20,16)

0.2

In [27]:
functionA(3,4,9)

0.15625

In [28]:
functionA(2,6,11)

0.027777777777777776

In [13]:
functionA(1,100,50)

0.5

### Function B

Determining the probability of rolling m dice, each with n sides, and having at least one die show a value greater than or equal to x.

In [29]:
def functionB (m, n, x):
  """
  Calculates the probability of rolling m dice, each with n sides, and having at least one die show a value greater than or equal to x.

  Args:
    m: The number of dice.
    n: The number of sides on each die.
    x: The target value.

  Returns:
    The probability of having at least one die show a value greater than or equal to x.
  """
  # Generate all possible rolls
  rolls = list(itertools.product(range(1, n+1), repeat=m))

  # Count the number of rolls with at least one die showing a value greater than or equal to x
  count = 0
  for roll in rolls:
    if any(die >= x for die in roll):
      count += 1

  # Calculate the probability
  return count / len(rolls)

**Test Cases**

In [30]:
functionB(2,6,1)

1.0

In [31]:
functionB(2,6,6)

0.3055555555555556

In [32]:
functionB(1, 100, 50)

0.51

### Evaluation

Evaluate the performance of your tool using the following scenario:

A player is in a challenging situation where they have a choice. They can choose between the following challenges:

• Roll 1d20 and get a result of 17 or more



We calculate this by having m = 1, n = 20, and x = 16. We use 16 instead of 17 because the specification of function A requires that we calculate the number of sums greater than x instead of greater than or equal to x.

In [22]:
functionA(1,20,16)

0.2

*Answer: 20%*

• Roll 3d4 and get a result of 10 or more



We calculate this by having m = 3, n = 4, and x = 9. We use 9 instead of 10 because the specification of function A requires that we calculate the number of sums greater than x instead of greater than or equal to x.

In [23]:
functionA(3,4,9)

0.15625

*Answer: 15.625%*

• Roll 4d6 and have at least one dice show a 6



We use functionB for this case, using m = 4, n = 6, and x = 6.

In [24]:
functionB(4,6,6)

0.5177469135802469

*Answer: 51.7745%*

Use your tool to compute the probabilities and report the best choice for the player.

**Conclusion: Based on the results of the functions it is in the best interest to use 4d6 with at least one 6.**